Index: tcp.py
===================================================================
--- tcp.py	(revision 13459)
+++ tcp.py	(working copy)
@@ -289,6 +289,7 @@
         self.socket.setblocking(0)
         self.fileno = skt.fileno
         self.protocol = protocol
+        self._readBuffer = '\0'*self.bufferSize
 
     if SSL:
 
@@ -339,13 +340,19 @@
         lost through an error in the physical recv(), this function will return
         the result of the dataReceived call.
         """
+        _readBuffer=self._readBuffer
+        self._readBuffer=None
+
         try:
-            data = self.socket.recv(self.bufferSize)
+            from crecv import recvinto
+            data = recvinto(self.socket.fileno(), self.bufferSize, _readBuffer)
         except socket.error, se:
+            self._readBuffer=_readBuffer
             if se.args[0] == EWOULDBLOCK:
                 return
             else:
                 return main.CONNECTION_LOST
+        self._readBuffer=_readBuffer
         if not data:
             return main.CONNECTION_DONE
         return self.protocol.dataReceived(data)
